webpack/components/Table/TableHooks.js

```diff
@@ -41,14 +42,151 @@ class ReactConnectedSet extends Set {
+
+  const areAllRowsSelected = () =>
+    Number(selectionSet.size) > 0 && selectionSet.size === Number(metadata.total);
+
+  const selectPage = () => selectionSet.addAll(ids);
+  const selectNone = () => selectionSet.clear();
+  const onRowSelect = (isSelected, id) => {
+    if (isSelected) {
+      selectionSet.add(id);
+    } else {
+      selectionSet.delete(id);
+    }
+  };
+
+  const selectedCount = selectionSet.size;
+
+  const isSelected = id => selectionSet.has(id);
+
+  return {
+    onRowSelect,
+    selectedCount,
+    areAllRowsOnPageSelected,
+    areAllRowsSelected,
+    selectPage,
+    selectNone,
+    isSelected,
+    selectionSet,
+  };
+};
+
+export const useBulkSelect = (results, metadata, initialArry = [], idColumn = 'id') => {
+  const { selectionSet: inclusionSet, ...selectOptions } =
+                useSelectionSet(results, metadata, initialArry, idColumn);
+  const exclusionSet = useSet([]);
+  const [searchQuery, updateSearchQuery] = useState('');
+  const [isSelectAllMode, setSelectAll] = useState(false);
+  const selectedCount = isSelectAllMode ?
+    Number(metadata.subtotal) - exclusionSet.size : selectOptions.selectedCount;
+
+  const areAllRowsOnPageSelected = () => isSelectAllMode ||
+                                         selectOptions.areAllRowsOnPageSelected();
+
+  const areAllRowsSelected = () => (isSelectAllMode && exclusionSet.size === 0) ||
+                                   selectOptions.areAllRowsSelected();
+
+  const isSelected = (id) => {
+    if (isSelectAllMode) {
+      return !exclusionSet.has(id);
+    }
+    return inclusionSet.has(id);
+  };
+
+  const selectPage = () => {
+    setSelectAll(false);
+    selectOptions.selectPage();
+  };
+
+  const selectNone = () => {
+    setSelectAll(false);
+    selectOptions.selectNone();
+  };
+
+  const onRowSelect = (isRowSelected, id) => {
+    if (isSelectAllMode) {
+      if (isRowSelected) {
+        exclusionSet.delete(id);
+      } else {
+        exclusionSet.add(id);
+      }
+    } else {
+      selectOptions.onRowSelect(isRowSelected, id);
+    }
+  };
+
+  const fetchBulkIds = () => {
+    const selected = {
+      included: {
+        ids: [],
+        search: null,
+      },
+      excluded: {
+        ids: [],
+      },
+    };
+
+    if (isSelectAllMode) {
+      selected.included.search = searchQuery;
+      selected.excluded.ids = [...exclusionSet];
+    } else if (!isEmpty(inclusionSet)) {
+      selected.included.ids = [...inclusionSet];
+    } else {
+      return {};
+    }
+    return selected;
+  };
+
+  const setSelectAllMode = (checked) => {
+    if (checked) {
+      exclusionSet.clear();
+    } else {
+      inclusionSet.clear();
+    }
+    setSelectAll(checked);
+  };
+
+  const clearSelections = () => {
+    setSelectAll(false);
+    inclusionSet.clear();
+    exclusionSet.clear();
+  };
+
+  return {
+    ...selectOptions,
+    selectPage,
+    selectNone,
```

Comment:
This should also expose a `selectAll` function.  No need to delegate or expose `selectAllMode` to the component-- just have the component call `selectAll` and handle it here.

Reply:
well I need the `isSelectAllMode` still for things like
```
  if (isSelectAllMode !== null) {
    selectAllDropdownItems.push(
      ( <DropdownItem key="select-all" component="button" isDisabled={isSelectAllMode} onClick={handleSelectAll}>
         {`${__('Select all')} (${totalCount})`}
        </DropdownItem>
      )
    );
  }

```

Reply:
From my reading it looks like you're using this check in `selectAllCheckbox` to determine if the `useBulkSelect` hook is in use, or just the regular `useSelectionSet`.

Seems simpler to me just to add a `showSelectAll` prop to `selectAllCheckbox` and use that.  I don't think `selectAllCheckbox` should need to know about `selectAllMode`. In fact, the only code that should know about it is TableHooks.js.

Reply:
Don't put anything in internal state which can be computed ;)

```js
isDisabled={areAllRowsSelected}
```